mediorum/server/repair.go

```diff
@@ -14,214 +13,214 @@ import (
-
-	// wait 10m after boot to start repair (plus up to 10 more minutes to space out nodes)
-	time.Sleep(time.Minute*10 + time.Minute*time.Duration(rand.Intn(10)))
-
+	// loop forever
 	for i := 1; ; i++ {
-		replicateMode := ss.shouldReplicate()
+		// wait between 10 and 30 minutes between repair runs
+		time.Sleep(time.Minute*10 + time.Minute*time.Duration(rand.Intn(20)))
+
 		// 10% percent of time... clean up over-replicated and pull under-replicated
-		cleanupMode := false
-		if i%10 == 0 {
-			cleanupMode = true
-		}
+		cleanupMode := i%10 == 0
 
 		logger := ss.logger.With("task", "repair", "run", i, "cleanupMode", cleanupMode)
-		took := time.Duration(0)
-		if replicateMode || cleanupMode {
-			repairStart := time.Now()
-			logger.Info("repair starting")
-			err := ss.runRepair(replicateMode, cleanupMode)
-			took = time.Since(repairStart)
-			if err != nil {
-				logger.Error("repair failed", "err", err, "took", took)
-			} else {
-				logger.Info("repair OK", "took", took)
-			}
-		} else {
+
+		// check that network is valid (should have more peers than replication factor)
+		if healthyPeers := ss.findHealthyPeers(5 * time.Minute); len(healthyPeers) < ss.Config.ReplicationFactor {
+			logger.Warn("not enough healthy peers to run repair",
+				"R", ss.Config.ReplicationFactor,
+				"peers", len(healthyPeers))
+			continue
+		}
+
+		// check that disk has space
+		if !ss.diskHasSpace() {
 			logger.Warn("disk has <200GB remaining and is not in cleanup mode. skipping repair")
+			continue
 		}
 
-		// sleep for as long as the job took
-		// if you wanted to aproximate max(1/5) of network running repair at same time...
-		// you could make this node run repair ~1/5 of the time
-		// (with some bounds)
-		sleep := clampDuration(time.Minute*5, took*5, time.Hour*12)
-		logger.Info("repair sleeping", "sleep", sleep)
-		time.Sleep(sleep)
+		repairStart := time.Now()
+		logger.Info("repair starting")
+		err := ss.runRepair(cleanupMode)
+		took := time.Since(repairStart)
+		if err != nil {
+			logger.Error("repair failed", "err", err, "took", took)
+		} else {
+			logger.Info("repair OK", "took", took)
+		}
 
 	}
 }
 
-func (ss *MediorumServer) runRepair(replicateMode, cleanupMode bool) error {
+func (ss *MediorumServer) runRepair(cleanupMode bool) error {
 	ctx := context.Background()
 
-	logger := ss.logger.With("task", "repair", "replicateMode", replicateMode, "cleanupMode", cleanupMode)
-
-	// check that network is valid (should have more peers than replication factor)
-	if healthyPeers := ss.findHealthyPeers(5 * time.Minute); len(healthyPeers) < ss.Config.ReplicationFactor {
-		return fmt.Errorf("invalid network: not enough healthy peers for R%d: %v", ss.Config.ReplicationFactor, healthyPeers)
+	// scroll uploads and repair CIDs
+	// (later this can clean up "derivative" images if we make image resizing dynamic)
+	for uploadCursor := ""; ; {
+		var uploads []Upload
+		ss.crud.DB.Where("id > ?", uploadCursor).Limit(5000).Find(&uploads)
```

Comment:
wouldn't this need an `order by`? also it seems reasonably fast now from a quick manual test, but afaik this `id` column isn't indexed so it could run into issues down the line. maybe it should get a `gorm:"index"` in the definition, or use OrigFileCID because it has an index already?

Reply:
`id` is the table primary key, so it is indexed... but yeah order by is totally necessary good catch!